### PR TITLE
[rllib] Add docs on how to use TF eager execution

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

Add docs on how to embed eager execution into custom models and loss functions. Unfortunately due to limitations of `tf.py_function` (namely, not allowing referencing non-eager tensors), we cannot transparently drop users into eager, but the examples here should be sufficient to show how to use it manually.

## Related issue number

Partially addresses https://github.com/ray-project/ray/issues/4921

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

File: doc/source/rllib-training.rst [hunk 1]
```diff
@@ -367,6 +367,13 @@ The ``"monitor": true`` config can be used to save Gym episode videos to the res
     openaigym.video.0.31403.video000000.meta.json
     openaigym.video.0.31403.video000000.mp4
 
+TensorFlow Eager
+~~~~~~~~~~~~~~~~
+
+While RLlib uses TF graph mode for all computations, you can still leverage TF eager to inspect the intermediate state of computations using `tf.py_function <https://www.tensorflow.org/api_docs/python/tf/py_function>`__. Here's an example of using eager mode in `a custom RLlib model and loss <https://github.com/ray-project/ray/blob/master/python/ray/rllib/examples/eager_execution.py>`__.
+
+There is also experimental support for running the entire loss function in eager mode. This can be enabled with ``use_eager: True``, e.g., ``rllib train --env=CartPole-v0 --run=PPO --config='{"use_eager": true, "simple_optimizer": true}'``. However this currently only works for a couple algorithms.
```
Comment (richardliaw) [inline doc/source/rllib-training.rst]: can you specify "a couple"

File: doc/source/rllib-concepts.rst [hunk 2]
```diff
@@ -346,6 +346,37 @@ In PPO we run ``setup_mixins`` before the loss function is called (i.e., ``befor
...
+        def compute_penalty(actions, rewards):
+            penalty = tf.reduce_mean(tf.cast(actions, tf.float32))
+            if random.random() > 0.9:
+                print("The eagerly computed penalty is", penalty, actions, rewards)
+            return penalty
+
+        actions = batch_tensors[SampleBatch.ACTIONS]
+        rewards = batch_tensors[SampleBatch.REWARDS]
+        penalty = tf.py_function(
+            compute_penalty, [actions, rewards], Tout=tf.float32)
```
Comment (richardliaw) [inline doc/source/rllib-concepts.rst]: So is the trick mainly to run py_functions in the static mode?

**Review by richardliaw — APPROVED** [review]

File: doc/source/rllib-training.rst (hunk 1 above)
Comment (ericl) [reply doc/source/rllib-training.rst]: Going to fill this out in ModelV2 -- hopefully by that time we will support most algos.

File: doc/source/rllib-concepts.rst (hunk 2 above)
Comment (ericl) [reply doc/source/rllib-concepts.rst]: Yep. Right now this is limited to loss_fn, but with ModelV2 this can also include the model forward pass.
